interface/ceed-preconditioning.c

```diff
@@ -800,99 +664,46 @@ static int CeedSingleOperatorAssemble(CeedOperator op, CeedInt offset,
-                B_mat_out[(qq+e_out)*elem_size + n] += interp_in[q * elem_size + n];
-              } else if (eval_mode_out[e_out] == CEED_EVAL_GRAD) {
-                d_out += 1;
-                B_mat_out[(qq+e_out)*elem_size + n] +=
-                  grad_in[(d_out*num_qpts+q) * elem_size + n];
-              } else {
-                // LCOV_EXCL_START
-                return CeedError(ceed, CEED_ERROR_UNSUPPORTED, "Not implemented!");
-                // LCOV_EXCL_STOP
-              }
-            }
-          }
-          for (CeedInt ei = 0; ei < num_eval_mode_out; ++ei) {
-            for (CeedInt ej = 0; ej < num_eval_mode_in; ++ej) {
-              const CeedInt eval_mode_index = ((ei*num_comp+comp_in)*num_eval_mode_in+ej)
-                                              *num_comp
-                                              +comp_out;
-              const CeedInt index = q*layout_qf[0] + eval_mode_index*layout_qf[1] +
-                                    e*layout_qf[2];
-              D_mat[(ei*num_eval_mode_in+ej)*num_qpts + q] += assembled_qf_array[index];
-            }
-          }
-        }
+  const CeedScalar qf_value_bound = max_norm*100*CEED_EPSILON;
+  for (CeedInt e = 0; e < num_elem; e++) {
+    for (CeedInt comp_in = 0; comp_in < num_comp; comp_in++) {
+      for (CeedInt comp_out = 0; comp_out < num_comp; comp_out++) {
         // Compute B^T*D
-        for (CeedInt ell = 0; ell < elem_size*num_qpts*num_eval_mode_in; ++ell) {
-          BTD[ell] = 0.0;
-        }
-        for (CeedInt j = 0; j<elem_size; ++j) {
-          for (CeedInt q = 0; q<num_qpts; ++q) {
-            const CeedInt qq = num_eval_mode_out*q;
-            for (CeedInt ei = 0; ei < num_eval_mode_in; ++ei) {
-              for (CeedInt ej = 0; ej < num_eval_mode_out; ++ej) {
-                BTD[j*(num_qpts*num_eval_mode_in) + (qq+ei)] +=
-                  B_mat_out[(qq+ej)*elem_size + j] * D_mat[(ei*num_eval_mode_in+ej)*num_qpts + q];
+        const CeedInt mat_size = elem_size * num_qpts*num_eval_mode_in;
+        for (CeedInt i = 0; i < mat_size; i++) BTD_mat[i] = 0.0;
+        for (CeedInt n = 0; n < elem_size; n++) {
+          for (CeedInt q = 0; q < num_qpts; q++) {
+            for (CeedInt e_in = 0; e_in < num_eval_mode_in; e_in++) {
+              for (CeedInt e_out = 0; e_out < num_eval_mode_out; e_out++) {
+                const CeedInt btd_index = n*(num_qpts*num_eval_mode_in) +
+                                          (num_eval_mode_in*q + e_in);
+                const CeedInt b_out_index = (num_eval_mode_out*q + e_out)*elem_size + n;
+                const CeedInt eval_mode_index = ((e_in*num_comp+comp_in)*num_eval_mode_out
+                                                 +e_out)*num_comp + comp_out;
+                const CeedInt qf_index = q*layout_qf[0] + eval_mode_index*layout_qf[1] +
+                                         e*layout_qf[2];
+                if (fabs(assembled_qf_array[qf_index]) > qf_value_bound) {
```

Comment:
I don't think we actually want this check here, do we?  We didn't have it previously in operator assembly (just diagonal assembly and FDMElementInverse).

Reply:
I don't see why not. If it is useful for diagonal assembly, its useful for full assembly as well. The idea was to reduce rounding noise from nearly zero entries in D.

Reply:
Well, for one thing, the other two routines (assemble diagonal and FDMElementInverse) are only meant for preconditioning purposes, no?  Whereas matrix assembly is billed as assembling the exact same operator as the "standard" matrix-free application.  We don't remove any noise from the QFunction application in that case.

Reply:
Diagonal assembly is billed as the exact, true diagonal of the assembled operator. Likely, it won't make a difference, so we can probably just strip in from both diagonal and full assembly.